Audio/SoundStreamInstance.cpp

```diff
@@ -0,0 +1,705 @@
+        if (mPackets[entry].state == State::FREE)
+        {
+            if (mCurrentPosition < mLengthInBytes)
+            {
+                auto cbValid = static_cast<uint32_t>(std::min(mPacketSize, mLengthInBytes - mCurrentPosition));
+
+                mPackets[entry].valid = cbValid;
+                mPackets[entry].audioBytes = 0;
+                mPackets[entry].startPosition = static_cast<uint32_t>(mCurrentPosition);
+                mPackets[entry].request.Offset = static_cast<DWORD>(mOffsetBytes + mCurrentPosition);
+
+                if (!ReadFile(async, mPackets[entry].buffer, uint32_t(mPacketSize), nullptr, &mPackets[entry].request))
+                {
+                    DWORD error = GetLastError();
+                    if (error != ERROR_IO_PENDING)
+                    {
+                        return HRESULT_FROM_WIN32(error);
+                    }
+                }
+
+                mCurrentPosition += cbValid;
+
+                mCurrentDiskReadBuffer = (entry + 1) % MAX_BUFFER_COUNT;
+
+                mPackets[entry].state = State::PENDING;
+
+                if ((cbValid < mPacketSize) && mLooped)
+                {
+#ifdef VERBOSE_TRACE
+                    DebugTrace("INFO (Streaming): Loop restart\n");
+#endif
+                    mCurrentPosition = 0;
+                }
+            }
+        }
+    }
+
+    return S_OK;
+}
+
+
+HRESULT SoundStreamInstance::Impl::PlayBuffers() noexcept
+{
+    HANDLE async = mWaveBank->GetAsyncHandle();
+
+    for (uint32_t j = 0; j < MAX_BUFFER_COUNT; ++j)
+    {
+        if (mPackets[j].state == State::PENDING)
+        {
+            DWORD cb = 0;
+#if (_WIN32_WINNT >= _WIN32_WINNT_WIN8)
+            BOOL result = GetOverlappedResultEx(async, &mPackets[j].request, &cb, 0, FALSE);
+#else
+            BOOL result = GetOverlappedResult(async, &mPackets[j].request, &cb, FALSE);
+#endif
+            if (result)
+            {
+                mPackets[j].state = State::READY;
+            }
+            else
+            {
+                DWORD error = GetLastError();
+                if (error != ERROR_IO_INCOMPLETE)
+                {
+                    ThrowIfFailed(HRESULT_FROM_WIN32(error));
+                }
+            }
+        }
+    }
+
+    if (!mBase.voice || !mPlaying)
+        return S_FALSE;
+
+    for (uint32_t j = 0; j < MAX_BUFFER_COUNT; ++j)
+    {
+        if (mPackets[mCurrentPlayBuffer].state != State::READY)
+            break;
+
+        const uint8_t* ptr = mPackets[mCurrentPlayBuffer].buffer;
+        uint32_t valid = mPackets[mCurrentPlayBuffer].valid;
+
+        bool endstream = false;
+        if (valid < mPacketSize)
+        {
+            endstream = true;
+#ifdef VERBOSE_TRACE
+            DebugTrace("INFO (Streaming): End of stream (%u of %zu bytes)\n", mPackets[mCurrentPlayBuffer].valid, mPacketSize);
+#endif
+        }
+
+        uint32_t thisFrameStitch = 0;
+        if (mStitchBytes > 0)
```

Comment:
PCM streaming is really easy and straight-forward. I wrote it years ago in [this sample](https://github.com/walbourn/directx-sdk-samples/tree/master/XAudio2/XAudio2AsyncStream).

ADPCM is much more complicated. The main issue is that MS-ADPCM has a block size that ranges from 32 to 512 and many of the block sizes are something like 70 bytes. That does *not* evenly divide into the required 2K page alignment required for non-buffering async read requests.

All of this stitch packet nonsense is handling blocks that straddle across the 2K page boundaries.